### PR TITLE
JNG-4594 transform crud behaviours

### PR DESCRIPTION
**Pull request by robertcsakany**
<table>
<td>
  <a href="https://blackbelt.atlassian.net/browse/JNG-4594" title="JNG-4594" target="_blank">Sub-taskJNG-4594</a>  Transform CRUD operations / behaviours
  </td></table>
  <br />

### CODE REVIEW

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/action/action.eol [hunk 1]
```diff
@@ -6,19 +6,60 @@ operation JSL!TransferActionDeclaration getId(): String {
...
+operation JSL!TransferActionDeclaration getOperationDeclarationEquivalent(): JUDOPSM!OperationDeclaration {
+
+	  if (self.eContainer.map.isUndefined()) {
+	  	return self.equivalent("CreateUnboundOperationForUnmappedTransferObjectType");	  
+	  } else if (self.eContainer.map.isDefined() and self.static) {
+  	  	return self.equivalent("CreateUnboundOperationForMappedTransferObjectType");	  	  
+	  } else if (self.eContainer.map.isDefined() and not self.static) {
+  	  	return self.equivalent("CreateBoundOperationForMappedTransferObjectType");	  	  
+	  } else {
+		throw "No choice defined";
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/action/action.eol]: Is exception name correct? I think it was copied from the function above.

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/data/transferDeleteDeclaration.eol [hunk 2]
```diff
@@ -0,0 +1,11 @@
+@cached
+operation JSL!TransferDeleteDeclaration getId(): String {
+    return (self.eContainer.getId() + "/" + self.name);
+}
+
+//@cached
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/data/transferDeleteDeclaration.eol]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/data/transferFieldDeclaration.eol [hunk 3]
```diff
@@ -17,3 +17,84 @@ operation JSL!TransferFieldDeclaration maps(): Boolean {
 operation JSL!TransferFieldDeclaration reads(): Boolean {
     return jslUtils.isReads(self);
 }
+
+@cached
+operation JSL!TransferFieldDeclaration getFqName(): String {
+    return self.eContainer.getFqName() + "#" + self.name;
+}
+
+
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/data/transferFieldDeclaration.eol]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/data/transferUpdateDeclaration.eol [hunk 4]
```diff
@@ -0,0 +1,11 @@
+@cached
+operation JSL!TransferUpdateDeclaration getId(): String {
+    return (self.eContainer.getId() + "/" + self.name);
+}
+
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/operations/jsl/data/transferUpdateDeclaration.eol:5]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/jslToPsm.etl [hunk 5]
```diff
@@ -24,6 +24,30 @@ import "modules/action/action.etl";
...
+
+import "modules/structure/transferDeclarationQueryCustomizer.etl";
+import "modules/actor/actorType.etl";
+import "modules/actor/access.etl";
+import "modules/action/accessListBehaviour.etl";
+import "modules/action/accessCreateBehaviour.etl";
+import "modules/action/accessValidateCreateBehaviour.etl";
+
 pre {
 	//var ecoreUtil = new Native("org.eclipse.emf.ecore.util.EcoreUtil");
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/jslToPsm.etl]: commented line

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/accessCreateBehaviour.etl [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+@lazy
+@greedy
+rule CreateCardinalityForCreateAccessInput
+    transform s : JSL!ActorAccessDeclaration
+    to t : JUDOPSM!Cardinality {
+      t.setId("(jsl/" + s.getId() + ")/CreateCardinalityForCreateAccessInput");
+
+      t.lower = 1; //s.isRequired and not s.isMany ? 1 : 0;
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/accessCreateBehaviour.etl]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/accessValidateCreateBehaviour.etl [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+@lazy
+@greedy
+rule CreateCardinalityForValidateCreateAccessInput
+    transform s : JSL!ActorAccessDeclaration
+    to t : JUDOPSM!Cardinality {
+      t.setId("(jsl/" + s.getId() + ")/CreateCardinalityForValidateCreateAccessInput");
+
+      t.lower = 1; //s.isRequired and not s.isMany ? 1 : 0;
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/accessValidateCreateBehaviour.etl:73]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/relationCreateBehaviour.etl [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+@lazy
+@greedy
+rule CreateCardinalityForCreateTransferTypeInput
+    transform s : JSL!TransferRelationDeclaration
+    to t : JUDOPSM!Cardinality {
+      t.setId("(jsl/" + s.getId() + ")/CreateCardinalityForCreateTransferTypeInput");
+
+      t.lower = 1; //s.isRequired and not s.isMany ? 1 : 0;
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/relationCreateBehaviour.etl]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/relationListBehaviour.etl [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    transform s : JSL!TransferRelationDeclaration
+    to t : JUDOPSM!Cardinality {
+      t.setId("(jsl/" + s.getId() + ")/CreateCardinalityForListEntityTypeOutput");
+
+      t.lower = s.isRequired and not s.isMany ? 1 : 0;
+      t.upper = s.isMany ? -1 : 1;
+}
+
+
+/*
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/action/relationListBehaviour.etl]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/actor/actorType.etl [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+      t.upper = 1;
+}
+
+
+rule CreateActorTypeClaim
+    transform s: JSL!ActorDeclaration
+    to t : JUDOPSM!TransferAttribute {
+        guard: s.getPrincipal().isDefined() 
+        	and s.getClaim().isDefined()
+        	and s.getIdentity().isDefined() // and s.getPrincipal().map.isDefined()
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/actor/actorType.etl]: commented code

File: judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationQueryCustomizer.etl [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+
+      t.setId("(jsl/" + s.getId() + ")/CreateQueryCustomizerSeekingTypeReverseAttribute");
+      t.name = "reverse";
+      t.required = false;
+      t.dataType = "extensions".equivalent("CreateQueryCustomizerBooleanType");
+      s.equivalent("CreateQueryCustomizerSeekingType").attributes.add(t);
+      log.debug("CreateQueryCustomizerSeekingTypeLimitAttribute type created: " + t.name);
+}
+
+
```
Comment (Madmat8) [inline judo-tatami-jsl-jsl2psm/src/main/epsilon/transformations/psm/modules/structure/transferDeclarationQueryCustomizer.etl]: commented codes

**Review by Madmat8 — APPROVED** [review]

**Review by gaborflorian — APPROVED** [review]

**Pull request merged**